open-telemetry/opentelemetry-python
# ext/jaeger: fix exporting to collector

Edit: I updated the PR to use THTTPClient directly

    The exporting of traces to the collector is broken, it replies with the
    "Unable to process request body: Required field Process is not set" error.
    
    The current implementation is based on OpenCensus [1], what appears to be broken
    too, it's not totally clear at this time what's wrong with that.
    
    This commit changes the logic to avoid using jaeger.Client that appears to be
    the broken piece, it is changed by a direct invokation to the THTTPClient
    transport from thrift.

    [1] https://github.com/census-instrumentation/opencensus-python/tree/master/contrib/opencensus-ext-jaeger


Fixes https://github.com/open-telemetry/opentelemetry-python/issues/493.

How to test:

Run the Jaeger collector, I used (notice there is not any port for the agent)
```
docker run -p 16686:16686 -p 14268:14268 jaegertracing/all-in-one
```

Export some traces to the collector:
```
# create a JaegerSpanExporter
jaeger_exporter = jaeger.JaegerSpanExporter(
    service_name="my-helloworld-service",
    # configure agent
    # agent_host_name="localhost",
    # agent_port=6831,
    # optional: configure also collector
    collector_host_name="localhost",
    collector_port=14268,
    collector_endpoint="/api/traces",
    # username="xxxx",  # optional
    # password="xxxx",  # optional
)

# create a BatchExportSpanProcessor and add the exporter to it
span_processor = BatchExportSpanProcessor(jaeger_exporter)

# add to the tracer factory
trace.get_tracer_provider().add_span_processor(span_processor)

# create some spans for testing
with tracer.start_as_current_span("foo") as foo:
    time.sleep(0.1)
    foo.set_attribute("my_atribbute", True)
    foo.add_event("event in foo", {"name": "foo1"})
    with tracer.start_as_current_span(
        "bar", links=[trace.Link(foo.get_context())]
    ) as bar:
        time.sleep(0.2)
        bar.set_attribute("speed", 100.0)

        with tracer.start_as_current_span("baz") as baz:
            time.sleep(0.3)
            baz.set_attribute("name", "mauricio")

        time.sleep(0.2)

    time.sleep(0.1)
```

Check that those traces where exported to Jaeger in http://localhost:16686.
